File: src/main/scala/org/tindalos/principle/domain/agents/sap/SAPResult.scala
package org.tindalos.principle.domain.agents.sap

import org.tindalos.principle.domain.agentscore.AnalysisResult
import org.tindalos.principle.domain.core.Package
import org.tindalos.principle.domain.expectations.SAP

import scala.collection.immutable.Seq

case class SAPResult(
    outlierPackages: Seq[Package],
    sapExpectation: SAP) extends AnalysisResult {

  val threshold = sapExpectation.violationsThreshold

  override def expectationsFailed() = outlierPackages.length > threshold
}

File: src/main/java/org/tindalos/principle/internal/domain/analyzers/sap/SAPResult.java
package org.tindalos.principle.internal.domain.analyzers.sap;

import org.tindalos.principle.internal.domain.core.AnalysisResult;
import org.tindalos.principle.internal.domain.constraints.SAP;
import org.tindalos.principle.internal.domain.core.packages.PackageWithMetrics;

import java.util.List;

/**
 * Represents the result of a Stable Abstractions Principle (SAP) analysis.
 * Contains packages whose distance from the main sequence exceeds the configured threshold.
 *
 * @param outlierPackages list of packages that violate the SAP constraint
 * @param sapExpectation  the SAP constraint configuration
 */
public record SAPResult(
        List<PackageWithMetrics> outlierPackages,
        SAP sapExpectation) implements AnalysisResult {

    public SAPResult {
        outlierPackages = List.copyOf(outlierPackages);
    }

    public int threshold() {
        return sapExpectation.violationThreshold();
    }

    @Override
    public boolean constraintViolated() {
        return outlierPackages.size() > threshold();
    }
}

